catroid/src/main/java/org/catrobat/catroid/utils/storagestuff/ProjectsStorageUtils.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.utils.storagestuff;
+
+import android.util.Log;
+
+import org.catrobat.catroid.common.FlavoredConstants;
+import org.catrobat.catroid.io.StorageOperations;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+
+import static org.catrobat.catroid.common.FlavoredConstants.EXTERNAL_STORAGE_ROOT_DIRECTORY;
+import static org.catrobat.catroid.common.FlavoredConstants.POCKET_CODE_EXTERNAL_STORAGE_FOLDER_NAME;
+import static org.catrobat.catroid.io.StorageOperations.getUniqueFile;
+import static org.catrobat.catroid.io.StorageOperations.transferData;
+
+public class ProjectsStorageUtils {
+
+	public static final String TAG = ProjectsStorageUtils.class.getSimpleName();
+	private static final String POCKET_CODE_INTERNAL_STORAGE_FOLDER_NAME = "files";
+	private static final String BACKPACK_FOLDER_NAME = "backpack";
+
+	private void copyFolderRecursive(File srcDir, File dstDir) throws IOException {
+		if (!srcDir.exists()) {
+			throw new FileNotFoundException("Directory: " + srcDir.getAbsolutePath() + " does not exist.");
+		}
+
+		if (!srcDir.isDirectory()) {
+			throw new IOException(srcDir.getAbsolutePath() + " is not a directory.");
+		}
+		if (!dstDir.exists()) {
+			dstDir.mkdir();
+		}
+
+		if (!dstDir.isDirectory()) {
+			throw new IOException("Cannot create directory: " + dstDir.getAbsolutePath());
+		}
+
+		for (File srcFile : srcDir.listFiles()) {
+			if (srcFile.isDirectory()) {
+				if (directoryContainsFile(srcFile, dstDir)) {
```

Comment:
the check of directoryContainsFile is unnecessary,

copyFolderRecursive(srcFile, getUniqueFile(srcFile.getName(), dstDir));

will just do fine for all, since it will return srcFile.getName if there is no file that it collides with.

Reply:
anyways, copyFolderRecursive should be replaced with copyDir from StorageOperations, then this method is not called recursively anymore. (then also a lot of the checks at the start of it could be omitted)